### PR TITLE
RPC: add weight to mempool entry output

### PR DESCRIPTION
**Pull request by esotericnonsense**
Tested against master using the REST api (/rest/mempool/contents), simple addition of a field.

Personal use case is for fee analysis software.

### CODE REVIEW

**Comment by promag** [comment]
No test affected 😞 care to improve by asserting the new field in the relevant RPC's 😉?

**Comment by laanwj** [comment]
Concept ACK

**Comment by MarcoFalke** [comment]
I guess you need to update the documentation as well.

**Comment by esotericnonsense** [comment]
Tests are failing.
txid2 and txid3 work as expected.
txid1 fails on both my test and https://github.com/bitcoin/bitcoin/pull/11203 (add wtxid to mempool entry output).

**Comment by ajtowns** [comment]
txid1 is failing because at that point "tx" is actually referring to the input to txid1, not the transaction for txid1. It works fine for me if I add

`         tx = FromHex(CTransaction(), self.nodes[0].gettransaction(txid1)['hex'])
`

prior to the assert_equal lines (and uncomment them obviously).

**Comment by esotericnonsense** [comment]
Doh. You're absolutely right. Fixed.

The final commit 'Refactor segwit 3-tx-chain' changes all references to 'tx' to 'tx/tx1/tx2/tx3' in order to clarify that. It has a large diff and can be dropped if necessary (only affects code style).

**Comment by morcos** [comment]
Concept ACK, but it turns out GetTxSize is not actually what we claim it is.  See the calculation of `GetVirtualTransactionSize` which potentially includes number of sig ops in the calculation.

I thinke the right path forward is: 
- a small BIP documenting this usage of virtual transaction size (which is used almost everywhere in our code
- documentation update pointing to this new BIP instead of 141 for defining virtual size
- this PR makes even more sense given this lack of easy conversion, but I think the newly added tests should at least be documented to note that they only work in the case that sig ops don't factor in, and perhaps we should add a test with a tx that has more sigops to show what happens in that case.

**Comment by luke-jr** [comment]
@morcos AFAIK that's exclusively used for node policy, and as such isn't a topic for standardisation...?

File: test/functional/segwit.py [hunk 1]
```diff
@@ -240,28 +240,45 @@ def run_test(self):
...
         assert(txid1 in self.nodes[0].getrawmempool())
+        tx1_hex = self.nodes[0].gettransaction(txid1)['hex']
+        tx1 = FromHex(CTransaction(), tx1_hex)
+
+        # Check that wtxid is properly reported in mempool entry (txid1)
+        assert_equal(int(self.nodes[0].getmempoolentry(txid1)["wtxid"], 16), tx1.calc_sha256(True))
+
+        # Check that weight and sizei (actually vsize) are properly reported in mempool entry (txid1)
+        assert_equal(self.nodes[0].getmempoolentry(txid1)["size"], (self.nodes[0].getmempoolentry(txid1)["weight"] + 3) // 4)
+        assert_equal(self.nodes[0].getmempoolentry(txid1)["weight"], len(tx1.serialize())*3 + len(tx1.serialize_with_witness()))
```
Comment (luke-jr) [inline test/functional/segwit.py]: serialize needs to be made serialize_without_witness

**Comment by MeshCollider** [comment]
Closing this, please see #14649

**Pull request closed without merging**